Add autoFlatMap

Jentsch opened the pull request:
Hi,
I think this library is very useful. Maybe it would be worthy to add `@autoMonad`.

The requirements for trait to have this annotations similar to `@autoFunktor`, only methods with the effect type or with a default implementation in the base trait are supported. This is a constraint from `pure`.

Open issues:

- [x] The current implementation isn't stack safe. That's probably fixable. 
- [x] I'm not sure about this copyright thingy, the new files are mostly a copy of the preexisting autoFunctor.scala
- [x] Map behaves not like the functor map, when the delegatee overrides a implementation of the base trait. See autoMonadTest Line 35.

kailuowang commented:
Thanks so much. I will have more time to take a thorough look tomorrow.

kailuowang commented on macros/src/main/scala/mainecoon/autoMonad.scala:
```diff
@@ -0,0 +1,76 @@
+              ..$pureMethods
+            }
+
+          override def tailRecM[A, B](a: A)(f: A => $name[..${tArgs(t"Either[A, B]")}]): $name[..${tArgs("B")}] =
```
I am curious if you've looked at the [monad implementation of `Function` in cats](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/instances/function.scala#L68-L76)? At a glance it might work to here make it stack-safe.

Jentsch replied:
I tried, but my first attempt was fruitless. But as soon, as I understood the idea and the types, it was quite easy. 96ddd76 fixes the issue.

Jentsch committed:
Remove unused method

Jentsch commented:
About the issue with the `map` function. Should we just enforce that every method without the effect type is final? I have no idea to come up with a lawful monad implementation.

kailuowang commented:
sorry I am not quite sure I understand why the `map` issue is happening. if the return type is not effect type, shouldn't it just delegate to the delegatee?

Jentsch commented:
No because map is implemented by using the pure method in cats.Monad:
```scala
override def map[A, B](fa: F[A])(f: A => B): F[B] =
    flatMap(fa)(a => pure(f(a)))
```
Sure map could be override but the underlying assumption (I think it's right identity) is still broken. The checkAll rule doesn't catch this, because SimpleAlg doesn't have such method.

I added a failing test to illustrate the problem.

Jentsch commented:
I rewrote the test so that checkAll finds the issue.

kailuowang commented:
I see. I wonder if you use the `map` generated from the `autoFunctor` will it pass the test?

kailuowang commented:
I suspect it would be hard to generate a lawful `pure` when there is overrides. So I think we can separate this into an `autoFlatMap` and an `autoMonad` (if you really need a `Monad` in your use case), the `autoMonad` would require all methods that come with implementation being final. (I think overrides on methods with effect type return type might break laws too, although I am not sure)

Jentsch commented:
No same test results, all identity properties fail.

Jentsch commented:
That's a good idea. I will investigate if monad has a use case. One big advantage (aside maybe existing use cases) is that they are very well known, `FlatMap` on the other side not so much. Currently I think the best way would be to add a check for non final methods and give a hint for `FlatMap` or `Functor` if making the methods final isn't an option.

About an `autoFlatMap`, I just tested if the current implementation satisfies the `FlatMap`-Laws, but that's apparently not the case:
```
FlatMap[TestAlg].flatMap.covariant identity *** FAILED ***
[...]
FlatMap[TestAlg].flatMap.invariant identity *** FAILED ***
```
But that should be fixable, since this test based upon the `map` function.

Finally methods which return the effect type should be for monads ok, since the result can be processed by the given function `f` in `flatMap(fa)(f)`.

Jentsch committed:
Unified matching for methods with different arguments list arity for functor

Jentsch committed:
Unified matching for methods with different arguments list arity for invariant

Jentsch committed:
Add handling of methods with an implementation in the base trait or class

Jentsch committed:
Add autoFlatMap

Jentsch commented:
Hi again,
basically this a rewrite of the first attempt. I had some trouble with fixing the ´map` function. Here is what changed:

- `@autoFunctor` now handles methods with default implementation
- the handling of such methods is checked by the `FunctorTests`
- the behavior is documented in the documentation
- added `@autoFlatMap` with an example in the documentation
- dropped `@autoMonad` for now (and renamed this PR)
- all copyrights follow now the same pattern: ` * Copyright 2017 Kailuo Wang`

kailuowang commented on tests/src/test/scala/mainecoon/tests/autoFunctorTests.scala:
```diff
@@ -18,55 +18,39 @@ package mainecoon
-    doubleAlg.foo("big") should be(3d)
-  }
+  checkAll("SimpleAlg[Option]", FunctorTests[TestAlgebra].functor[Long, String, Int])
+  checkAll("Functor[SimpleAlg]", SerializableTests.serializable(Functor[TestAlgebra]))
```
since you replaced `SimpleAlg` with `TestAlgebra`, would you change the text here as well?

kailuowang commented on tests/src/test/scala/mainecoon/tests/autoFlatMapTests.scala:
```diff
@@ -0,0 +1,94 @@
+class autoFlatMapTests extends MainecoonTestSuite {
+
+  checkAll("FlatMap[TestAlg]", FlatMapTests[TestAlgebra].flatMap[Float, String, Int])
+  checkAll("FlatMap[SimpleAlg]", SerializableTests.serializable(FlatMap[TestAlgebra]))
```
since you replaced `SimpleAlg` with `TestAlgebra`, would you change the text here as well?

kailuowang commented on tests/src/test/scala/mainecoon/tests/autoFunctorTests.scala:
```diff
@@ -18,55 +18,39 @@ package mainecoon
 
-  @autoFunctor
-  trait AlgWithDef[T] {
-    def foo: T
```
just noticed that this is not tested before, but at least it compiled. Can we add a def method without arguments in `TestAlgebra` to test this scenario as well?

Jentsch replied:
Done

kailuowang commented on macros/src/main/scala/mainecoon/autoFlatMap.scala:
```diff
@@ -0,0 +1,140 @@
+    import cls._
+
+
+    val mapMethods = templ.stats.toList.flatMap(_.collect {
```
The duplication of this mapMethods code between `autoFlatMap` and `autoFunctor` is a pity. Is there a straightforward way to reduce it a little bit? (maybe a method that takes a `ad: AlgDefn` and the mapFunction name and delegatee name?) what do you think? What do you think, at the very minimum we can leave a comment here suggesting that it's duplicated with `autoFunctor`.

Jentsch replied:
I created a common `mapMethods` function in autoFlatMap.

Jentsch committed:
reuse common map function of flatMap and functor

kailuowang commented on tests/src/test/scala/mainecoon/tests/autoFlatMapTests.scala:
```diff
@@ -0,0 +1,99 @@
+      (p.abstractEffect: String => T) ***
+        (p.concreteEffect: String => T) ***
+        (p.abstractOther: String => String) ***
+        (p.concreteOther: String => String)
```
shall we add `withoutParameters` here too?

Jentsch replied:
I was aware how to add here `withoutParameters` with the arrow operation. But it was total unnecessary, simple tuppeling works also.

kailuowang commented on macros/src/main/scala/mainecoon/autoFlatMap.scala:
```diff
@@ -134,6 +115,27 @@ object autoFlatMap {
 
   }
+
+  private[mainecoon] def mapMethods(template: Template): List[Defn.Def] = {
```
build failed. we probably need to pass in `effectTypeName` as an argument

Jentsch replied:
You're right, don't know how this slipped through.

Jentsch commented:
@kailuowang Thanks for all your feedback.

kailuowang commented:
Thank you for this amazing contribution!
